### ISSUE TITLE
Disagreement with missing_panics_doc changes regarding `expect`

### ISSUE THREAD

**Issue by ecton** [rust-lang/rust-clippy#11436]
### Summary

In #10953, `missing_panics_doc` was updated to include reporting on `Option::expect` and `Result::expect`. This has caused a large number of warnings across my projects, and I wanted to start a discussion over the changes.

To me, these two approaches to asserting that an option shouldn't be `None` are identical from a programmer's intent:

```rust
pub fn example(opt: Option<()>) {
    match opt {
        Some(_) => {}
        None => unreachable!("invalid"),
    }

    opt.expect("invalid")
}
```

Despite the first `match` statement possibly panicking, no warning is given.

The net result of this change is that I'm going to be forced into an ugly pattern in my code to avoid this false positive. Could we split the warnings for panicking on `expect` into its own lint so that we can keep the previous behavior?

### Lint Name

missing_panics_doc

### Reproducer

I tried this code:

```rust
pub fn example(opt: Option<()>) {
    opt.expect("invalid")
}
```

I saw this happen:

```
warning: docs for function which may panic missing `# Panics` section
   --> src/shapes.rs:442:5
    |
442 |     pub fn test(i: Option<()>) {
    |     ^^^^^^^^^^^^^^^^^^^^^^^^^^
    |
note: first possible panic found here
   --> src/shapes.rs:448:9
    |
448 |         i.expect("invalid");
    |         ^^^^^^^^^^^^^^^^^^^
    = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#missing_panics_doc
```

I expected to see this happen: No warnings

### Version

```text
rustc 1.72.0 (5680fa18f 2023-08-23)
binary: rustc
commit-hash: 5680fa18feaa87f3ff04063800aec256c3d4b4be
commit-date: 2023-08-23
host: x86_64-unknown-linux-gnu
release: 1.72.0
LLVM version: 16.0.5
```

**Comment by PKilcommons** [comment]
I have similar reservations about this change so I figure I'd chime in.

Recently saw it show up in CI and was a bit surprised. While my line of thought on this matter may be potentially unidiomatic, I consider usages of `.except()` to be similar to `// SAFETY` comments for `unsafe` blocks in the body of a safe function—dev-to-dev explanation of an invariant being upheld in a way the compile cannot infer. A panic is then a result of that invariant being broken inside the function body, not as a result of the input to the function. As such this is then a bug that needs to be fixed *within* the body of the function, not at the callsite. 

In this case, the `# Panics` section ends up just becoming pointless noise in the doc comment simply repeating the message from the `.except()`, something the caller has zero control over.

To illustrate with a contrived function:

```rust
/// Return a `String` pulled out of the byte stream.
pub fn string_from_byte_stream() -> String {
    let bytes = get_valid_utf8();
    String::from_utf8(bytes).except("`get_valid_utf8()` always returns valid UTF-8")
}

fn get_valid_utf8() -> Vec<u8> {
    /* Gets some bytes that the dev knows is always valid UTF-8 */
}
```
 
To appease the lint now:
```rust
/// Return a `String` pulled out of the byte stream.
/// # Panics
/// Panics if the byte stream is not valid UTF-8.
pub fn string_from_byte_stream() -> String {
    /* ... */ 
}
```
 
This doesn't exactly help the caller prevent or avoid a panic in anyway.

Perhaps adding a `clippy.toml` config variable that allows the `.except()` case would also be a desirable alternative to a new lint?

**Comment by oxalica** [comment]
Running into this recently. I use `.expect()` and/or `.unwrap()` interchangebly  for already-checked-conditions or invariants that another library/code guarantee. So if it really panics, it's my fault or upstream library's fault. It's meaningless for downstream users to know in documentations that it may panic, because the end users cannot control or fix it at all.

cc the author of that change @KisaragiEffective

**Comment by Alexendoo** [comment]
A list of methods/macro names to ignore seems like a good config option to add. There have been several different incompatible conventions mentioned for what `unwrap`/`expect` mean so that seems like the best way forward

**Comment by ModProg** [comment]
@Alexendoo 
> A list of methods/macro names to ignore seems like a good config option to add.

What should the config be? How should it handle paths, e.g. `std::*`, `core::*`, 

Initial idea, though one could also consider an allowlist instead:
```toml
panics_requiring_doc = ["Option::unwrap", "Option::expect", "Result::unwrap", "Result::expect", "panic!", "assert!", "assert_eq!", "assert_ne!"]
```

One issue is that the detection is implemented quite differently:

https://github.com/rust-lang/rust-clippy/blob/789bc73d8a7068ff2ec658db9e8ec197f86d0954/clippy_utils/src/macros.rs#L194-L207

https://github.com/rust-lang/rust-clippy/blob/789bc73d8a7068ff2ec658db9e8ec197f86d0954/clippy_lints/src/doc.rs#L843-L846

and

https://github.com/rust-lang/rust-clippy/blob/789bc73d8a7068ff2ec658db9e8ec197f86d0954/clippy_lints/src/doc.rs#L852-L856

**Comment by Alexendoo** [comment]
Personally I would go with a list of paths to ignore as that's more common for clippy configuration, you could take inspiration from `disallowed_macros`/`disallowed_methods`

**Comment by ModProg** [comment]
> Personally I would go with a list of paths to ignore as that's more common for clippy configuration, you could take inspiration from `disallowed_macros`/`disallowed_methods`

my idea was that one could add their own functions/macros to the lint, but personally I don't have a usecase for that, so just having an ignore list is fine by me

**Comment by duelafn** [comment]
My deepest wish is that I could flag individual lines/blocks as not requiring panic docs. I haven't looked at the lint code and fully expect that what I want might be difficult to impossible, but in the world of dreams, I'd really like to be able to (reusing some above examples),

``` rust
/// Do something
pub fn string_from_byte_stream() -> String {
    let bytes = get_valid_utf8();
    #[expect(clippy::missing_panics_doc_ok, reason="caller can't do anything about this")]
    String::from_utf8(bytes).expect("`get_valid_utf8()` always returns valid UTF-8")
}
```

Time passes, function modified,

``` rust
/// Do something
pub fn string_from_byte_stream(input: i64) -> String {
    let input = u64::try_from(input).expect("non-negative");// Oops, added something that needs "# Panics"

    // ... stuff

    let bytes = get_valid_utf8();
    #[expect(clippy::missing_panics_doc_ok, reason="caller can't do anything about this")]
    String::from_utf8(bytes).expect("`get_valid_utf8()` always returns valid UTF-8")
}
```

In my dreams the first fn wouldn't lint, but the second one would since it has an untagged panic.

**Comment by KisaragiEffective** [comment]
you can use the _unchecked variant to eliminate both warning and unnecessary control flow jump, aren't you?

**Comment by KisaragiEffective** [comment]
Aside that, can we rename this as `unexplained_panic` and restore old behavior to exclude the `expect` method?

**Comment by ModProg** [comment]
I started using a crate (https://docs.rs/intentional/latest/intentional/trait.Assert.html) that provides some helper methods that do the same as expect & unwrap so I can differentiate between cases I want to document and cases I don't.

**Comment by Alexendoo** [comment]
Opened https://github.com/rust-lang/rust-clippy/pull/14407 for `allow`/`expect` within functions

For `assert`/`panic` etc it would have to be on a parent expression as attributes directly on macro calls are currently ignored, which may mean an extra block has to be introduced

```rust
pub fn do_something(input: &[i64]) {
    if handle_small_vec(input) {
        return;
    }

    #[expect(clippy::missing_panics_doc_ok)]
    {
        assert!(
            input.len() > 3,
            "small vecs handled above, this assert satisfies clippy::missing_asserts_for_indexing"
        );
    }
    do_something_else((input[0] - input[1]) * (input[2] - input[3]));
}
```
